File: openbas-api/src/main/java/io/openbas/utils/pagination/PaginationField.java
package io.openbas.utils.pagination;

import io.swagger.v3.oas.annotations.media.Schema;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.springframework.data.domain.Sort;

import java.util.ArrayList;
import java.util.List;

@AllArgsConstructor
@NoArgsConstructor
@Builder
@Data
public class PaginationField {

  @Schema(description = "Filter object to search within filterable attributes")
  private FilterGroup filterGroup;

  @Schema(description = "Text to search within searchable attributes")
  private String textSearch;

  @Schema(description = "List of sort fields : a field is composed of a property (for instance \"label\" and an optional direction (\"asc\" is assumed if no direction is specified) : (\"desc\", \"asc\")")
  private List<SortField> sorts = new ArrayList<>();

  public Sort getSort() {
    List<Sort.Order> orders;

    if (null == this.sorts || this.sorts.isEmpty()) {
      orders = List.of();
    } else {
      orders = this.sorts.stream()
          .map(field -> {
            String property = field.property();
            Sort.Direction direction = Sort.DEFAULT_DIRECTION;
            if (null != field.direction()) {
              String directionString = field.direction();
              direction = Sort.Direction.fromOptionalString(directionString).orElse(Sort.DEFAULT_DIRECTION);
            }
            return new Sort.Order(direction, property);
          }).toList();
    }

    return Sort.by(orders);
  }

  public enum FilterMode {
    and,
    or
  }

  public enum FilterOperator {
    eq
  }

  @Data
  public static class FilterGroup {
    private FilterMode mode; // Between filters
    private List<Filter> filters;
  }

  @Data
  public static class Filter {
    private String key;
    private FilterMode mode; // Between values: name = name1 OR name = name2
    private List<String> values;
    private FilterOperator operator;
  }
}

File: openbas-front/src/components/common/pagination/PaginationField.ts
export interface PaginationField {
  filterGroup?: FilterGroup;
  textSearch?: string;
  sorts?: SortField[];
}

enum FilterOperator {
  eq = 'eq',
}

enum FilterMode {
  and = 'and',
  or = 'or',
}

interface FilterGroup {
  mode: FilterMode;
  filters?: Filter[];
}

export interface Filter {
  key: string;
  mode: FilterMode;
  values: string[];
  operator: FilterOperator;
}

interface SortField {
  property?: string;
  direction?: string;
}

export const initFilterGroup: (key: string, values: string[]) => FilterGroup = (key: string, values: string[]) => {
  return {
    mode: FilterMode.and,
    filters: [{
      key,
      values,
      mode: FilterMode.and,
      operator: FilterOperator.eq,
    }],
  };
};

export const initSorting: (property: string) => SortField[] = (property: string) => {
  return [{ property }];
};

// Pageable

export interface Page<T> {
  content: T[];
  empty: boolean;
  first: boolean;
  last: boolean;
  number: number;
  numberOfElements: number;
  pageable: Pageable;
  size: number;
  sort: Sort;
  totalElements: number;
  totalPages: number;
}

interface Pageable {
  offset: number;
  pageNumber: number;
  pageSize: number;
  paged: boolean;
  sort: Sort;
  unpaged: boolean;
}

interface Sort {
  empty: boolean;
  sorted: boolean;
  unsorted: boolean;
}
